### PR TITLE
Pull wireless from config, enabling use in image and initramfs

### LINKED ISSUE #17: configure wifi on boot from config
right now the boot and the device work on the wired network interface (eth0) only. 
this would add support to configure the wifi / enable network connectivity via wifi. 
on 1st boot the device would need to be on the wired network, get the wifi params and after that it could work without a wired connection.

### PR DESCRIPTION
**Pull request by zmrow**
Once testing is completed, this should fix #17 

After a few iterations and much pondering, I believe this is the safest, simplest, and easiest way to make wireless connectivity possible both in the rasbian image as well as the initramfs.

Basically if the `.wpa_supplicant` key exists in the config, we decode it and put it in the `${CATTLEPI_DIR}/wpa_supplicant.conf`.  This makes it available in subsequent boots to both the initramfs and (after injecting it) the image.  We also attempt to update the `wpa_supplicant.conf` every boot.

Documentation for [wpa_suppicant is here](https://www.raspberrypi.org/documentation/configuration/wireless/wireless-cli.md).

I'm happy to further explain why I went about it this way, but honestly this is pretty straight forward and I couldn't see how my other (more-complex) methods would have been any better.

### CODE REVIEW

File: templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers [hunk 1]
```diff
@@ -352,12 +352,42 @@ cattlepi_build_root_filesystem()
...
+        fi
+    else
+        mv /tmp/wpa_supplicant.conf "${CATTLEPI_DIR}/wpa_supplicant.conf"
+        echo "wpa_supplicant.conf created; this will be used next boot"
+    fi
+}
+
 cattlepi_bring_up_wlan() {
     interface="${1:-wlan0}"
+    wpa_supplicant_location="${2:-/etc/wpa_supplicant/wpa_supplicant.conf}"
```
Comment (mirceal) [inline templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers]: should you use this var?

File: templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers [hunk 2]
```diff
@@ -352,12 +352,42 @@ cattlepi_build_root_filesystem()
...
+    if [ $? -ne 0 ]; then
+        echo "Invalid or empty wpa_supplicant, skipping wireless"
+        return
+    fi
+
+    if [ -r "${CATTLEPI_DIR}/wpa_supplicant.conf" ]; then
+        cmp -s "${CATTLEPI_DIR}/wpa_supplicant.conf" /tmp/wpa_supplicant.conf
+        if [ $? -ne 0 ]; then
+            mv /tmp/wpa_supplicant.conf "${CATTLEPI_DIR}/wpa_supplicant.conf"
+            echo "wpa_supplicant.conf updated"
```
Comment (mirceal) [inline templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers]: should we force a reboot if the supplicant was written? this way we could use it immediately. 
also, one question i had was: if the initrd writes the supplicant, can the actual raspbian os in which we're booting use ti the first time? (or do we need a reboot also?)

File: templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers (hunk 1 above)
Comment (zmrow) [reply templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers]: Oops - missed that.  Will do!

File: templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers (hunk 2 above)
Comment (zmrow) [reply templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers]: Hmm.  Yeah, I suppose we could reboot right away?

After wpa_supplicant is written to $CATTLEPI_DIR in this step, later in the boot process the build step runs which should copy the file to the raspbian filesystem in the appropriate place.

**Comment by zmrow** [comment]
That's awesome!! Can' wait to see the e2e

**Comment by zmrow** [comment]
Oh.... Looks like this will require a change to the API as well?  Docs say that keys will get ignored unless they are known.  I'm guessing the API doesn't know about a key "wpa_supplicant".

**Comment by mirceal** [comment]
Yep.
The api does not actually care about what you put in the configuration json. It’s the different bits and pieces in the image that give meaning to the config. 
So, yes you can add stuff in there, and as long as it’s still a valid json we should be good to go.

**Comment by zmrow** [comment]
Update - I finally had a bit of time to dig into why my `POST`s weren't working.  I was trying to POST to the API (`make localapi_run`) to update the config with `wpa_supplicant`,  but was getting a 405.  

Derpaderp the local API server doesn't seem support that method yet!!  I have a quickly hacked together file-based JSON API fix and should be able to run through this before the weekend.

Apologies for the long wait.

**Comment by zmrow** [comment]
Working through a few issues I found, will update this PR once I fix them.

**Comment by zmrow** [comment]
The file is correctly created and put in CATTLEPI_DIR as well as /etc/wpa_supplicant.

In `netup` we check whether the wpa_supplicant file is readable and then use it.  However even though the file exists and should be readable, that block of code in `netup` being skipped.  Really not sure why this is happening.
😕 😖

**Comment by mirceal** [comment]
The netup runs before the rootfs is built. So when it’s running there is nothing there. Later in the process it the rootfs is built but by this time it’s too late? (This is just a guess)

**Comment by zmrow** [comment]
Correct, rootfs won't be available.  I just realized that `/dev/mmcblk0p1` won't be mounted to `/boot` at the time `netup` runs.  Since we use /boot/cattlepi as persistent storage for things like `config` and `wpa_supplicant`, it must be available in order to use the `wpa_supplicant` for wireless during initramfs runtime.  Of course, if nothing is there, we won't use anything.

Is there any reason you can see not to mount `/boot` during `netup`?

**Comment by mirceal** [comment]
Somewhat agree w/ all of the above. I do NOT want to alter the initramfs as this would take us on an interesting path. 

Here is what I would suggest: let’s add a few methods to the helpers that deal with mounting [ro and rw] and unmounting the sdcard. Invoke those where it’s appropriate. Remove the sdmount and sdmountro steps. 

I would also strongly recommend a reboot if the supplicant is updated. (It’s okay if the pi reboots once or twice while booting and picking up “persisting” changes - this ensures it’s going to work if it loses power while up and comes back w/ no network). I am already doing this if I need to partition the sd card.

File: templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers [hunk 3]
```diff
@@ -447,3 +447,22 @@ cattlepi_bring_up_wlan() {
...
 	fi
 }
+
+cattlepi_sdmount() {
+    # Explicitly mount the sdcard partition to /boot
+    # We don't give the caller any other option as of right
+    # now since this partition doesn't need to be mounted
+    # anywhere else and doing so could be dangerous
+    echo "Mounting sdcard to /boot"
+    mkdir /boot
```
Comment (mirceal) [inline templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers]: you probably need the unmount  here too

File: templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers [hunk 4]
```diff
@@ -447,3 +447,22 @@ cattlepi_bring_up_wlan() {
...
+    mount -n -t vfat /dev/mmcblk0p1 /boot
+}
+
+cattlepi_sdmount_RO() {
+    # Explicitly mount the sdcard patition RO to /boot
+    # As above, don't give the caller any other option
+    # since we don't need to and it could be dangerous yadda yadda
+    echo "Mounting sdcard RO to /boot"
+    umount /boot
```
Comment (mirceal) [inline templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers]: add mkdir here too

File: templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers [hunk 5]
```diff
@@ -447,3 +447,22 @@ cattlepi_bring_up_wlan() {
...
+    # Explicitly mount the sdcard partition to /boot
+    # We don't give the caller any other option as of right
+    # now since this partition doesn't need to be mounted
+    # anywhere else and doing so could be dangerous
+    echo "Mounting sdcard to /boot"
+    mkdir /boot
+    mount -n -t vfat /dev/mmcblk0p1 /boot
+}
+
+cattlepi_sdmount_RO() {
```
Comment (mirceal) [inline templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers]: probably need at cattlepi_umount function that can be invoked. 
also i would prefer cattlepi_sdmount_ro but don't have a very strong opinion on it

File: templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers (hunk 4 above)
Comment (zmrow) [reply templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers]: 👍

File: templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers (hunk 3 above)
Comment (zmrow) [reply templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers]: 👍

File: templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers (hunk 5 above)
Comment (zmrow) [reply templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers]: 👍

File: templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers [hunk 6]
```diff
@@ -414,3 +447,24 @@ cattlepi_bring_up_wlan() {
...
+    mount -n -t vfat /dev/mmcblk0p1 /boot
+}
+
+cattlepi_sdmount_ro() {
+    # Explicitly mount the sdcard patition RO to /boot
+    # As above, don't give the caller any other option
+    # since we don't need to and it could be dangerous yadda yadda
+    echo "Mounting sdcard RO to /boot"
+    mkdir /boot
+    umount /boot
```
Comment (mirceal) [inline templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers]: insert a **sync** after this umount

File: templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers [hunk 7]
```diff
@@ -414,3 +447,24 @@ cattlepi_bring_up_wlan() {
...
 }
+
+cattlepi_sdmount() {
+    # Explicitly mount the sdcard partition to /boot
+    # We don't give the caller any other option as of right
+    # now since this partition doesn't need to be mounted
+    # anywhere else and doing so could be dangerous
+    echo "Mounting sdcard to /boot"
+    mkdir /boot
+    umount /boot
```
Comment (mirceal) [inline templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers]: insert a **sync** after this umount

File: templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers (hunk 6 above)
Comment (mirceal) [reply templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers]: again, what I think happens is that you mount the SD card, you do a bunch of downloads / writes to it, but when you switch it to readonly the kernel does not finish writing everything to card (vfat behavior). You end up with partial/corrupted files that fail the md5 check. 

I saw something similar when I was playing with the sdlayout. see here: https://github.com/cattlepi/cattlepi/blob/34a2f6696eb9e6b40a9acb34e5e9d4a7cc840d1e/templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers#L245

you also probably want to sync whever you are using the sysreq-trigger, as for example here: https://github.com/cattlepi/cattlepi/blob/d70a7f512bd4c8ea8aa3bdf68d7fce2741f4be96/templates/raspbian_cattlepi/resources/bin/autoupdate.sh#L19

File: templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers (hunk 7 above)
Comment (zmrow) [reply templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers]: Wouldn't we want to sync the data to disk *before* the umount? Or am I thinking about that wrong?

Comment (mirceal) [reply templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers]: Huh. Now that you explicitely ask about it i'm not that sure. 
On a "normal" filesystem (ie not vfat) the umount should flush all the pending writes to disk. Normally if you copy a large file to the card you would expect that upon unmount the kernel would wait for the writes to complete. What I have observed is that it does not and you end up with a corrupted file.

Now, Should the sync be before or after the umount. Techincally I think the right answer is before the unmount (but after any copy/write operation). You force everything to be written and when the sync completes the umount is a formality.
Why does it work this way? (after umount). I think when I was researching this I read somewhere that the umount for vfat is done in a lazy manner. What this means is that the kernel does the bookkeeping and after that the writes will be flushed out (eventually?? saw this by having it sleep for tens of seconds). So the sync after the umount actually forces it to do the writes now.
So lazy umount + sync == proper umount. 

I have a hard time finding the reference where the lazy behavior for vfat was documented now. I recall that after I did this it seemed to work each time so I never though about it too much (which is why I think it's pretty cool when someone else can verify/question your though process :+1: )

So to answer your question: it probably works both ways and it should probably be _before_ the umount. 
No, you are not thinkign about it wrong. Any sane filesystem would not need the sync to umount properly.

Comment (zmrow) [reply templates/raspbian_cattlepi/resources/usr/share/initramfs-tools/scripts/cattlepi-base/helpers]: Having all sorts of issues trying to get the bcrmfmac driver to load properly and the wlan0 interface to show up.  Still investigating and will continue to work on this.

**Comment by zmrow** [comment]
This is fully working now!  You can have a fully wireless `raspbian_cattlepi` image after the initial boot and setup.  This means only need to boot the Pi hardwired to the network once and after that it can function fully wirelessly.  (I have opened up issue #91 to track the documentation of this feature)

You need a few things for this to function correctly.
1. A valid `wpa_supplicant.conf` file that is base64 encoded (just as we do with the usercode).  An example of said file is below. See the [Raspberry Pi docs here](https://www.raspberrypi.org/documentation/configuration/wireless/wireless-cli.md) for more info.
2. This base64 encoded file must live in the `config` under the top-level key `wpa_supplicant`.  (At the same level as `usercode`, etc.

Even with proper wireless config, a fully booted raspbian_cattlepi machine that has an ethernet cable plugged into it will give precedence to the hardwired network.  I personally feel that this is a good default.  Feel free to pull the cable, reboot the box (just to be safe) and your machine can function wirelessly from then on forward.

Example `wpa_supplicant`:
```
ctrl_interface=DIR=/var/run/wpa_supplicant GROUP=netdev
update_config=1
country=US

network={
        ssid="cool_wireless_ssid"
        psk="REDACTED"
        key_mgmt=WPA-PSK
}
```
Example `config`:
```
{
  "usercode": "",
  "wpa_supplicant": "Awholebunchofbase64encodeddata",
  "bootcode": "",
...
}
```


* Removed the "WIP" from the title
* Opened issue #91 to document how this works and how to set it up.

**Comment by mirceal** [comment]
thank you

**Pull request merged**